### PR TITLE
Go HTTP-only part 3

### PR DESCRIPTION
**Pull request by trustin**
Modifications:

- Simplify the type parameter signatures of services and clients even
  further
  - There's no point of using `? super I, ? extends O` now that we
    expect them to be `RpcRequest, RpcResponse` or `HttpRequest, HttpResponse`.
- Remove `SessionProtocolProvider`
- Deprecate `HttpSessionProtocols`
- Make `SessionProtocol` an enum and undeprecate its values
- Remove the redundant `if (HttpSessionProtocols.isHttp(...))`
- Make `RequestLog` and `RequestLogBuilder` more HTTP-friendly
  - Replace `requestEnvelope()` with `requestHeaders()`
  - Replace `responseEnvelope()` with `responseHeaders()`
  - Replace `RequestLogAvailability.REQUEST_ENVELOPE` with `REQUEST_HEADERS`
  - Replace `RequestLogAvailability.RESPONSE_ENVELOPE` with `RESPONSE_HEADERS`
  - Add `status()` which is `HttpStatus`
  - Remove `RequestLogAvailability.STATUS_CODE` because `HttpStatus`
    becomes available when `responseEnvelope` becomes available

Result:

- Even more simplicity
- More HTTP-friendliness

To-dos:

- Relocate some classes
- Replace the `ClientRequestContext.HTTP_HEADERS` attribute with a
  first-class property of `ClientRequestContext`

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/SessionProtocol.java [hunk 1]
```diff
@@ -17,93 +17,53 @@
...
+    H2C("h2c", false, true, 80);
 
-    /**
-     * @deprecated Use {@link HttpSessionProtocols#values()}.
-     */
-    @Deprecated
-    public static Set<SessionProtocol> ofHttp() {
-        return HttpSessionProtocols.values();
-    }
+    private static final BiMap<String, SessionProtocol> uriTextToProtocols;
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/common/SessionProtocol.java]: I don't see any usages of the reversed map - do you have future plans for using this `BiMap`?

File: core/src/main/java/com/linecorp/armeria/common/http/HttpSessionProtocols.java [hunk 2]
```diff
@@ -18,61 +18,71 @@
...
+import java.util.EnumSet;
 import java.util.Set;
 
-import com.google.common.collect.ImmutableSet;
-
 import com.linecorp.armeria.common.SessionProtocol;
 
 /**
- * HTTP-related {@link SessionProtocol} instances.
+ * @deprecated Use {@link SessionProtocol} instead.
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/common/http/HttpSessionProtocols.java]: Hooray! Was getting annoyed by IDE not suggesting these properly.

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestLog.java [hunk 3]
```diff
@@ -266,19 +267,29 @@ default void ensureAvailability(Iterable<RequestLogAvailability> properties) {
     String query();
 
     /**
-     * Returns the status code specific to the {@link Response} of the current {@link SessionProtocol}.
+     * Returns the status of the {@link Response}.
      *
      * @throws RequestLogAvailabilityException if this property is not available yet
      */
-    int statusCode();
+    default HttpStatus status() {
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestLog.java]: `@Nullable`

File: core/src/main/java/com/linecorp/armeria/common/SessionProtocol.java (hunk 1 above)
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/SessionProtocol.java]: Probably not. Let me fix it. :-)

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestLog.java (hunk 3 above)
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestLog.java]: Reviewed the nullability of all methods and parameters and added `@Nullable` to the eligible params and methods.

**Comment by trustin** [comment]
Thank you, reviewers!

**Pull request merged**